File: src/main/scala/com/vc/hard/L45.scala
package com.vc.hard

object L45 {
  def jump(arr: Array[Int]): Int = {
    var curEnd = 0
    var curFarthest = 0
    var jumps = 0
    (0 until arr.length - 1).foreach(i => {
      curFarthest = Math.max(i + arr(i), curFarthest)
      if(i == curEnd) {
        jumps += 1
        curEnd = curFarthest
      }
    })
    jumps
  }
}

File: src/main/java/com/vc/hard/L45.java
package com.vc.hard;

import java.util.*;

class L45 {
    public int jump(int[] arr) {
        int n = arr.length;
        int[] res = new int[n];
        Arrays.fill(res, Integer.MAX_VALUE);
        for(int i = 0; i < n - 1; i++) {
            int k = 1;
            while(k <= arr[i] && k + i < n) {
                res[k + i] = res[i] == Integer.MAX_VALUE ? 1 : Math.min(res[k + i], res[i] + 1);
                k++;
            }
        }
        res[0] = 0;
        return res[n - 1];
    }
}
